clang/lib/Driver/Driver.cpp

```diff
@@ -2489,6 +2497,181 @@ void Driver::BuildInputs(const ToolChain &TC, DerivedArgList &Args,
+  // actually archives.
+  std::vector<StringRef> BundlerArgs = {"clang-offload-bundler", "-type=ao",
+                                        Targets, Inputs, "-check-section"};
+  return runBundler(BundlerArgs, C);
+}
+
+// Simple helper function for Linker options, where the option is valid if
+// it has '-' or '--' as the designator.
+static bool optionMatches(const std::string &Option,
+                          const std::string &OptCheck) {
+  return (Option == OptCheck || ("-" + Option) == OptCheck);
+}
+
+// Process linker inputs for use with offload static libraries.  We are only
+// handling options and explicitly named static archives as these need to be
+// partially linked.
+static void getLinkerArgs(Compilation &C, DerivedArgList &Args,
+                          SmallVector<const char *, 16> &LibArgs) {
+  for (const auto *A : Args) {
+    std::string FileName = A->getAsString(Args);
+    if (A->getOption().getKind() == Option::InputClass) {
+      std::string Value(A->getValue());
+      if (isStaticArchiveFile(Value)) {
+        LibArgs.push_back(Args.MakeArgString(FileName));
+        continue;
+      }
+    }
+    if (A->getOption().hasFlag(options::LinkerInput)) {
+      // Do not add any libraries that are not fully named static libs
+      if (A->getOption().matches(options::OPT_l) ||
+          A->getOption().matches(options::OPT_reserved_lib_Group) ||
+          A->getOption().hasFlag(options::NoArgumentUnused))
+        continue;
+      for (const std::string &Value : A->getValues()) {
+        static std::string PrevArg;
+        auto addKnownValues = [&](const std::string &V) {
+          // Only add named static libs objects and --whole-archive options.
+          if (optionMatches("-whole-archive", V) ||
+              optionMatches("-no-whole-archive", V) || isStaticArchiveFile(V)) {
+            LibArgs.push_back(Args.MakeArgString(V));
+            return;
+          }
+          // Probably not the best way to handle this, but there are options
+          // that take arguments which we should not add to the known values.
+          // Handle -z for now - can be expanded if/when usage shows the need
+          if (PrevArg != "-z" && V[0] != '-' && isObjectFile(V)) {
+            LibArgs.push_back(Args.MakeArgString(V));
+            return;
+          }
+        };
+        if (Value[0] == '@') {
+          // Found a response file, we want to expand contents to try and
+          // discover more libraries and options.
+          SmallVector<const char *, 20> ExpandArgs;
+          ExpandArgs.push_back(Value.c_str());
+
+          llvm::BumpPtrAllocator A;
+          llvm::StringSaver S(A);
+          llvm::cl::ExpandResponseFiles(
+              S,
+              C.getDefaultToolChain().getTriple().isWindowsMSVCEnvironment()
+                  ? llvm::cl::TokenizeWindowsCommandLine
+                  : llvm::cl::TokenizeGNUCommandLine,
+              ExpandArgs);
+          for (std::string MA : ExpandArgs)
```

Comment:
I'd suggest using `StringRef` whenever possible. Also, `MA` is a tough name to grasp - could you please explain or change it?

Reply:
I have a bad habit of using `MA` and not using better values.  I also made adjustments to use StringRef a bit more.